Add Quick-Access to favorite folder in left sidepanel in files-app #9720

skjnldsv opened the pull request:
Somehow broke #9720 (sorry @newhinton) :(

Please review @nextcloud/designers 
Already a :+1: from me!

Commit:
Added FavoritesQuickaccess-Sidebar

Added FavoritesQuickaccess-Sidebar

Added Files-FavoritesQuickaccess-Toggle

Fixed CSS for SpacerElement

Removed Unnessessary Alerts and added Translations

Tried fixing initial Quick-Access Checkboxstate

Signed-off-by: fnuesse <fnuesse@techfak.uni-bielefeld.de>

Tried fixing initial Quick-Access Checkboxstate

Changed double-Quotes to single-Quotes

Revert webdavurl which was changed by mistake

Revert quota-icon which was changed by mistake

Changed the Folderhandling from custom-designed to nextcloud-NavigationManager-handling

Signed-off-by: fnuesse <fnuesse@techfak.uni-bielefeld.de>

Moved CSS-Spacerclass to apps.scss for global usage

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Renamed settings-caption in apps.scss to app-navigation-caption

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Removed old input-tag for showQuickAccess-state

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Removed old spacer element in files.scss

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Changed style of favorites-sublist and disabled the ability to disable files-quickaccess

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

newhinton committed:
Added toggleable QuickAccess

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Deleted wrongly commited File

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Added quickaccess-state persistence

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Started implementing sorting-algorithm for proper alphabetical display

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Finished Sorting of Quickaccess elements

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Implemented persistence for reverse-list and sorting-strategy

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Implemented initial sorting for reverse-list and sorting-strategy

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Refactored Code

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Refactored Code

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Refactored Code

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

newhinton committed:
Fixed complete app-breakdown if no favorites are set

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Hid three-dot menu if no favorites are selected

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

newhinton committed:
Added responsive folderadding

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Added collapsible to quickaccesslist

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Refactored Code

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

newhinton committed:
Fixed css-flashing

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed menu if initially empty

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

newhinton committed:
Changed toggle

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed Codestyle and changed linkgeneration

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed Codestyle

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed Codestyle (Renamed Variables)

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

newhinton committed:
Added copyright notice

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

newhinton committed:
Removed old code

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Removed duplicate collapse-button and changed api-endpoints

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Removed app-navigation-caption from apps.scss

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Changed api-endpoints

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed Codestyle (.js)

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Hid away extended Settings

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed reverse state

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed Missing reverse after changing sort-strategy

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed Copyright-Header

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Removed UI-Flickering

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

hid dotmenu on toggle while favorites are empty

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Added Draggable to listelements (WIP)

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Rebuild appnavigation.php with recursive function to allow easy implementation of sublists

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed draggable Sublist-Elements

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed draggable Sublist-Elements

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Added date-modified sorting option to quickaccess

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Added custom order sorting option to quickaccess

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Added custom order sorting option to quickaccess

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Added fallback for custom ordering

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

newhinton committed:
Set Quickaccess-ordering to custom order

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

newhinton committed:
Fix collapsible-Button when list was emptied (or got first element)

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed Bad url-generation in javascript for new quickaccessitems

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed vertical scrolling in sortable-list which leads to "hidden" navbar

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Removed unnessessary console logs

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed Bounds in custom sorting

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Reformatted code

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed horizontalscroll on sortable-list
Fixed "stuck element" where you could not switch back to the original ordering in the sortable-list

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

newhinton committed:
Added drop-animation to sortable-list

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Added icon-change on drag

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed Navbar-closing in app when favorites-list is toggled on mobile

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Refactored Code

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Changed to alphabetical sorting

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed deletion of folder with identical names

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Removed ability to add files to the quickaccess

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Fixed wrong path-generation when added from favorites-star

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Removed Element from navbar when favorite-star in detailview is toggled off

Signed-off-by: fnuesse <felix.nuesse@t-online.de>

Changed Quota-Text to prevent boundarybreaks

Reverted last commit

juliusknorr reviewed (dismissed):
- [ ] Toggling the quick access list is not possible
- [x] I think we should hide it by default, otherwise some users who already have a bunch of favorites could end up with a long list of folders there and the other sections would be hidden on the first run
- [ ] Adding a folder to the favorites will not update the list in the sidebar
- [x] Clicking the folder flashes a second indicator, which looks a bit weird.
![image](https://user-images.githubusercontent.com/3404133/42512948-17eef1c0-8456-11e8-8197-e3160c70bc70.png)

- [x] There is also a JS error when clicking a folder in the sidebar:
``` js
Uncaught Error: Syntax error, unrecognized expression: li[data-id=http://localhost:8140/index.php/apps/files/?dir=/4 ]
    at Function.ga.error (core.js?v=d502b209-3:2)
    at ga.tokenize (core.js?v=d502b209-3:2)
    at Function.ga [as find] (core.js?v=d502b209-3:2)
    at a.fn.init.find (core.js?v=d502b209-3:2)
    at a.fn.init.a.fn.find (core.js?v=d502b209-3:7)
    at Navigation.setActiveItem (merged-index.js?v=d502b209-3:11396)
    at Navigation._onClickItem (merged-index.js?v=d502b209-3:11421)
    at HTMLDivElement.dispatch (core.js?v=d502b209-3:3)
    at HTMLDivElement.r.handle (core.js?v=d502b209-3:3)
```

juliushaertl commented:
@newhinton Besides those comments, I really like it. 

@skjnldsv @newhinton Drag-and-drop is not included in the PR, right?

skjnldsv commented:
>> Toggling the quick access list is not possible
>> Clicking the folder flashes a second indicator, which looks a bit weird.
>
> I cant reproduce those

I can, I'll fix it :)

> At the moment it is only disabled. The code is still there, because i thougt that the "how to sort"-discussion wasn't over yet. I have the idea of an dropdown menu, which allows changing of the sorting-order, but i wanted a statement for that because that would introduce a dropdown-menu to the ui, which needs to be discussed. If you want to take a peak, look at newhinton:WithSettingsDropdown. I have tested it there.

@newhinton you can push to this pr!
Nonetheless, let's not add a dropdown, @jancborchardt is not fully on board for now, so let’s just sort them by last modified up top. :)

juliushaertl commented:
@newhinton Chrome 67

>>Adding a folder to the favorites will not update the list in the sidebar
> Do you mean the ordering is not updated?

The folder that is added to the favorites is missing in the sidebar until i reload the whole page.

newhinton commented:
@juliushaertl Sorry, i was confused there, drag&drop is not implemented at all. What i meant was the custom order, which is ordered by drag&drop :D

@skjnldsv thanks for fixing it!

skjnldsv committed:
Fix app nav menu highlight

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Default state to collapsed and fixed code styling

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented on apps/files/lib/Controller/ViewController.php:
```diff
@@ -159,28 +160,76 @@ public function index($dir = '', $view = '', $fileid = null, $fileNotFound = fal
+		foreach ($favElements['folders'] as $elem) {
+
+			$id = substr($elem, strrpos($elem, '/') + 1, strlen($elem));
+			$link = $this->urlGenerator->linkToRouteAbsolute('files.view.index', ['dir' => $elem]);
```
Somehow the link generated here does not redirect to the proper view in fies :/
@juliushaertl  any clue?

skjnldsv replied:
clicking this redirect from `/apps/files/?dir=/Test`
to `/apps/files/dir=/&view=Test`

skjnldsv committed:
Fixed invalid uid

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Fixed favorite folder view display

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

MorrisJobke reviewed (approved):
Tested and works 👍

MorrisJobke commented:
Unit test failure:

```
1) OCA\Files\Tests\Controller\ViewControllerTest::testIndexWithRegularBrowser
Your test case is not allowed to access the database.
```

skjnldsv commented:
> Unit test failure:

Any clue? I'm on the jsunit atm :)

MorrisJobke commented on apps/files/lib/Controller/ViewController.php:
```diff
@@ -159,28 +160,77 @@ public function index($dir = '', $view = '', $fileid = null, $fileNotFound = fal
+		$helper = new \OCA\Files\Activity\Helper($tagger);
+
+		try {
+			$favElements = $helper->getFavoriteFilePaths($this->userSession->getUser()->getUID());
```
This one is likely to cause the DB used exception. This one is called in a test that has no DB but actually needs one.

rullzer replied:
I think I fixed the tests

rullzer committed:
Fix tests

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Inject helper

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

juliusknorr reviewed (approved):
Works great for me now :+1: 

Good to go, if the jsunit tests are fixed. :wink:

skjnldsv committed:
Fixed tests

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
All fixed, let's wait for final tests
https://drone.nextcloud.com/nextcloud/server/8883

rullzer committed:
No draggable list anymore

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
phpunit test broken! fix pushed

newhinton commented:
@skjnldsv I see you have removed the _setOnDrag function, i get that, but why do you keep all the other stuff, which is now dead code?

skjnldsv commented:
@newhinton I removed it because the drag wasn't working properly :)
I thought I cleaned all the related stuff, could you point out where there are still stuff left please?

Thanks for your amazing work, this pr is a very nice addition to nc14!!

newhinton commented:
@skjnldsv I would prefer to fix it instead of removing it, what do you say?

newhinton commented:
In that case i meant the sortable handler (for custom sorting). While we are not using it for this feature, the handling of different sorting strategies is fully working in the current files-app, and the only thing currently breaking sortablelist is the missing handler.
 When we would readd the handler, it would work again (after fixing your mentioned bugs).

This enables us to use this feature later, when (and if) we found a solution for the ui-problem (look at [newhinton:WithSettingsDropdown](https://github.com/newhinton/server/tree/WithSettingsDropdown) , i have experimented there) without redeveloping everything.
Also, slight changes could make the draggablething dynamic for multiple lists, like the upcoming pr for the shared entries, as an example.

If you want to remove it, (i can remove it, if you whish) you need to delete the api-routes, you need to remove the unnessessary sorting-variable in navigation.js, edit the quicksort-helper function to only return the alphabetical sorting value and edit the php-code to dump the initial-sorting.

I believe my solution for sorting is quiet dynamic there, and while not perfect, provides a lot of possibilities and i would keep it, but in the end, it's your decision ;)

skjnldsv commented:
No, we don't want to add a menu of any sort in the end. Let keep things simple, the ui already start to be a bit overcrowded :)

Please, go ahead for the cleanup, I'm on another issue ;)
Thanks a lot!

newhinton commented:
@skjnldsv How do i push the cleanup? Should i open a new PR?

skjnldsv commented:
Yes, on nextcloud/server please :)
